keps/sig-architecture/20191023-conformance-without-beta.md

```diff
@@ -0,0 +1,169 @@
+* Kubernetes components requiring behavior enabled by beta feature gates
+* Kubernetes conformance tests requiring calls to beta REST APIs
+* Kubernetes conformance tests exercising behavior enabled by beta feature gates
+
+For each category, we should identify and resolve existing dependencies, and prevent future occurrences.
+
+### Goals
+
+* Identify existing beta REST APIs and features required by Kubernetes components so they can be graduated or the dependencies removed
+* Identify existing beta REST APIs and features required by conformance tests so the tests can be rewritten to remove those dependencies
+* Prevent new beta REST APIs and features from being required by Kubernetes components or conformance tests
+* Demonstrate passing conformance tests on a cluster with all beta REST APIs and features disabled
+
+### Non-Goals
+
+* Resolve questions about conformance profiles (see https://github.com/kubernetes/community/issues/2651)
+* Resolve questions about GA-but-optional features (see https://github.com/kubernetes/community/issues/3997)
+* Resolve code-level dependencies in Kubernetes source code on pre-v1.0.0 components
+* Forbid beta REST APIs or features from being enabled in conformant clusters by distributors
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes component developer, I cannot accidentally make changes
+that would require a production cluster to enable a beta API or feature
+in order to pass conformance tests.
+
+Attempts to make such a change would be caught/blocked by a CI job.
+
+#### Story 2
+
+As a Kubernetes conformance test author, I cannot accidentally write 
+(or promote to conformance) tests that require use of beta REST APIs.
+
+Attempts to make such a change would be caught/blocked by a CI job.
+
+#### Story 3
+
+As a Kubernetes distribution maintainer, I can pass conformance tests
+without enabling beta REST APIs or features.
+
+A canonical CI job running against Kubernetes master and release branches
+would demonstrate that it is possible to pass conformance tests with no
+beta REST APIs or features enabled.
+
+## Design Details
+
+1. Make it possible to easily disable all beta REST APIs and features
+  * Add support for disabling REST API versions of the form `v[0-9]+beta[0-9]+` to the Kubernetes API server with `--runtime-config=api/beta=false`
+    * Parallels use of `--runtime-config=api/all=false`
+    * For completeness, also add support for disabling REST API versions of the form `v[0-9]+alpha[0-9]+` to the Kubernetes API server with `--runtime-config=api/alpha=false`
+  * Add support for disabling beta feature gates to all components with `--feature-gates=AllBeta=false`
+    * Parallels use of `--feature-gates=AllAlpha=false`
+2. Identify existing beta REST APIs/features required by Kubernetes components or conformance tests
```

Comment:
Here as the 14 Conformance tests that currently hit alpha or beta api operations directly:

```sql-mode
select distinct
   split_part(useragent, '--', 2) as test
from audit_event
where job = '1183553991464718336'
  and useragent like 'e2e.test%Conformance%'
  and (operation_id ilike '%alpha%' or
       operation_id ilike '%beta%');
```

```sql-mode
                                                                              test                                                                               
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
  [k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart http hook properly [NodeConformance] [Conformance]
  [k8s.io] Container Runtime blackbox test when starting a container that exits should run with the expected status [NodeConformance] [Conformance]
  [k8s.io] InitContainer [NodeConformance] should invoke init containers on a RestartNever pod [Conformance]
  [k8s.io] InitContainer [NodeConformance] should not start app containers and fail the pod if init containers fail on a RestartNever pod [Conformance]
  [k8s.io] Kubelet when scheduling a busybox Pod with hostAliases should write entries to /etc/hosts [LinuxOnly] [NodeConformance] [Conformance]
  [k8s.io] Kubelet when scheduling a read only busybox container should not write to root filesystem [LinuxOnly] [NodeConformance] [Conformance]
  [k8s.io] Pods should support retrieving logs from the container over websockets [NodeConformance] [Conformance]
  [k8s.io] Probing container should be restarted with a exec "cat /tmp/health" liveness probe [NodeConformance] [Conformance]
  [k8s.io] Security Context When creating a container with runAsUser should run the container with uid 65534 [LinuxOnly] [NodeConformance] [Conformance]
  [sig-api-machinery] Aggregator Should be able to support the 1.10 Sample API Server using the current Aggregator [Conformance]
  [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] works for multiple CRDs of different groups [Conformance]
  [sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] works for multiple CRDs of same group and version but different kinds [Conformance]
  [sig-apps] Job should run a job to completion when tasks sometimes fail and are locally restarted [Conformance]
  [sig-network] DNS should provide /etc/hosts entries for the cluster [LinuxOnly] [Conformance]
(14 rows)

```

Reply:
We can look a bit closer at each test as required to create a burndown of beta/alpha operations that need removal in order to clean them up, or demote them until they can be rewritten.

Reply:
#### The 3 alpha operation hit by some of the above 14 Conformance tests
```sql-mode
select distinct
  operation_id
from audit_event
where job = '1183553991464718336'
  and useragent like 'e2e.test%Conformance%'
  and operation_id ilike '%alpha%'
order by operation_id;
```

```sql-mode
           operation_id            
-----------------------------------
 getBatchV2alpha1APIResources
 getSchedulingV1alpha1APIResources
 getSettingsV1alpha1APIResources
(3 rows)

```

Reply:
#### 21 beta apis
```sql-mode
select distinct
  operation_id,
  object_group as group,
  object_ver as version
  -- object_kind as kind
from audit_event
where job = '1183553991464718336'
  and useragent like 'e2e.test%Conformance%'
  and operation_id ilike '%beta%'
order by operation_id;
```

```sql-mode
                operation_id                 | group  | version 
---------------------------------------------+--------+---------
 createPolicyV1beta1PodSecurityPolicy        | policy | v1beta1
 getAdmissionregistrationV1beta1APIResources |        | 
 getApiextensionsV1beta1APIResources         |        | 
 getApiregistrationV1beta1APIResources       |        | 
 getAuthenticationV1beta1APIResources        |        | 
 getAuthorizationV1beta1APIResources         |        | 
 getAutoscalingV2beta1APIResources           |        | 
 getAutoscalingV2beta2APIResources           |        | 
 getBatchV1beta1APIResources                 |        | 
 getCertificatesV1beta1APIResources          |        | 
 getCoordinationV1beta1APIResources          |        | 
 getEventsV1beta1APIResources                |        | 
 getExtensionsV1beta1APIResources            |        | 
 getNetworkingV1beta1APIResources            |        | 
 getNodeV1beta1APIResources                  |        | 
 getPolicyV1beta1APIResources                |        | 
 getRbacAuthorizationV1beta1APIResources     |        | 
 getSchedulingV1beta1APIResources            |        | 
 getStorageV1beta1APIResources               |        | 
 listPolicyV1beta1PodSecurityPolicy          | policy | v1beta1
 readPolicyV1beta1PodSecurityPolicy          | policy | v1beta1
(21 rows)

```

Reply:
Thanks for the data. I'm fairly confident some of the `get` requests are false-positives, due to things that enumerate all served APIs and fetch them. Running conformance tests with those APIs turned off completely would clarify which were specifically used/required by the tests and components.